Fix: Account for comments in implicit-arrow-linebreak

peanutenthusiast opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[X ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

`Implicit-arrow-linebreak` does not take comments into consideration when the option for the rule is `beside`.

When the rule evaluates the expression:
```
(foo) =>
   //comment
   bar;
```

The outcome is:
```
(foo) => bar;
```

The proposed fix is:
```
// comment
(foo) => bar;
```

**What changes did you make? (Give an overview)**
- Add test cases including comments in `tests/lib/rules/implicit-arrow-linebreak`
- Add helper for formatting comment text split by line breaks
- Implement filtering condition, and fix for when the above bug occurs

**Is there anything you'd like reviewers to focus on?**
- This is my first PR for this project -- any feedback would be heavily appreciated.
- Do the test cases suffice? Any missing logic?
- What should be the `outcome` with the test case of
```
(foo) =>
  // hi
     bar =>
       // there
         braz;
```
? This one was a doozy.
@sharmilajesupaul

ilyavolodin commented:
Hmm... It seems like this fix is improving current behavior, but I'm not 100% sure this is the best way. In the last example you showed, the only reasonable way to fix it based on your approach would be to move all of the comments above the line:
```js
// hi
// there
(foo) =>bar => braz;
```

However, chances are, if there's a chaining like that, comments describe what each step of the chain is doing, and if you move them above, it would be unclear. I think best way to fix that would be to change line comments into inline comments.

```js
(foo) => /* hi */ bar => /* there */ braz;
```

But that might not work and might possibly change the meaning of the code. So I'm not sure what would be the best way to do that.

not-an-aardvark commented:
I think the best solution would be to just not do an autofix if a line comment is found in that location. We have a similar behavior with `brace-style`, where we don't move the opening bracket when the code looks like this:

```js
if (foo) // comment
{
}
```

ljharb commented:
I'd expect:
```js
(foo) =>
  // hi
     bar =>
       // there
         braz;
```

to become:
```js
(foo) => (
  // hi
     bar => (
       // there
         braz
      )
);
```

peanutenthusiast commented:
@ljharb @not-an-aardvark @ilyavolodin Thanks for all of your input. It seems like prepending the arrow functions with parentheses clarifies the intent of the comments.

I will go ahead and implement this change.

sharmilajesupaul commented on lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -72,7 +92,27 @@ module.exports = {
+                        if (comments.length > 0) {
+                            const firstToken = sourceCode.getFirstToken(
+                                node.parent.type === "VariableDeclarator"
+                                    ? node.parent.parent : node
```
What is `node.parent.parent` in this case?

sharmilajesupaul commented on lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -4,6 +4,8 @@
  */
 "use strict";
 
+const astUtils = require("../ast-utils");
```
looks like this is unused

sharmilajesupaul commented on lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -72,7 +92,27 @@ module.exports = {
+                    fix: fixer => {
+                        const placeBesides = fixer.replaceTextRange([tokenBefore.range[1], fixerTarget.range[0]], " ");
+
+                        if (comments.length > 0) {
```
nit: this can be `if(comments.length)`

peanutenthusiast commented:
@sharmilajesupaul Thanks for the review and comments! Still working on the above test case as well as when:
```
async (foo) => 
    // comment
    bar
```

sharmilajesupaul commented:
@peanutenthusiast instead of trying to find the parent and placing the comment before it. What if you got the line number from the node range and placed the comment on the line before that with a prepended `\n`?

Also, does this account for cases where you are in an object, or a method chain, or array? 

eg: 

```js
const foo = {
  id: 'bar'
  prop: (foo1) =>
    // comment
    'returning this string', 
}

[ foo => 
  // comment
  'bar'
]

"foo".split('').map((char) => 
  // comment
  `-${char}-`
))
```

can you add test cases for those? ^

peanutenthusiast commented:
@sharmilajesupaul will need to account for the aforementioned test cases as well. thanks!

peanutenthusiast commented:
@sharmilajesupaul Would the expected output for the method chain test case be
```
// comment
"foo".split('').map((char) => `-${char}-`
)
```
or (probably nicer on the eyes)
```
// comment
"foo".split('').map((char) => `-${char}-`)
```
?

sharmilajesupaul commented:
@peanutenthusiast The paren linebreak is handled by [`function-paren-newline`](https://eslint.org/docs/rules/function-paren-newline) so I don't think you need to worry about includin that in your autofix. 

There might be cases where you do want the paren on it's own line:
```js
"foo".split('').map(
  (char) => 
    `-${char}-`,
)
```
I think that might be out of the scope of this PR.

ljharb commented:
this seems like it needs a fresh rebase

peanutenthusiast commented:
@ljharb agreed and working on it.

peanutenthusiast commented:
Any updates with this PR? If any sharp eyes notice test cases that still need accounting for, more than happy to take care of them.

sharmilajesupaul commented on lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -37,6 +43,127 @@ module.exports = {
+
+                if (comments.length) {
+
+                    const precedingArrow = sourceCode.getTokenBefore(arrow, isArrowToken);
```
I'm not sure what this is doing, maybe add a comment explaining this?
why do you return null if there is a preceding arrow?

peanutenthusiast replied:
`arrowToken` would be a better name instead of `arrow`.

`addParentheses()` appends the necessary parentheses by checking if there is a following arrow, so if a `precedingArrow` exists, that logic has already been handled, so I return null.

sharmilajesupaul commented on tests/lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -132,6 +132,218 @@ ruleTester.run("implicit-arrow-linebreak", rule, {
+        }, {
+            code: `
+                (foo) =>
+                  // test comment
```
maybe add valid examples for these test cases with comments?

sharmilajesupaul commented on tests/lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -132,6 +132,218 @@ ruleTester.run("implicit-arrow-linebreak", rule, {
+        }, {
+            code: `
+                (foo) =>
+                  /* test comment */
```
as we discussed, add more test cases with multiline block comments

sharmilajesupaul commented:
hey @peanutenthusiast! is this ready for another review?

peanutenthusiast commented:
Yes!

On Tue, Dec 4, 2018 at 5:13 PM Sharmila Jesupaul <notifications@github.com>
wrote:

> hey @peanutenthusiast <https://github.com/peanutenthusiast>! is this
> ready for another review?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/10545#issuecomment-444320455>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Aa-ISTkZKKdXwI4v_Ay3_rbsHsKqe_e7ks5u1x2vgaJpZM4U9zaM>
> .
>

platinumazure commented on tests/lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -132,6 +199,317 @@ ruleTester.run("implicit-arrow-linebreak", rule, {
+                    // comment
+                    // another comment
+                        bar`,
+            output: "// comment\n// another comment\n(foo) => bar",
```
Nit: It would really help to see this represented as a template since the code example is also a template-- that way it's easier to see the differences visually.

peanutenthusiast replied:
@platinumazure Thanks for the comment! I applied the requested change to this test case and the one below.

platinumazure replied:
Thanks @peanutenthusiast, much appreciated! I'll review again tonight.

nzakas commented:
@peanutenthusiast Sorry we lost track of this!

@platinumazure have your changes been addressed?

@ilyavolodin @not-an-aardvark thoughts on moving this forward?

not-an-aardvark commented:
Should this be marked as a bug rather than an enhancement? Fixers generally shouldn't be deleting comments.

platinumazure commented:
This PR is labeled as "enhancement" but the commit summary says "Fix". Is this a bug fix or an enhancement?

not-an-aardvark commented:
I'm going to re-label it as a bug -- I think the original labelling was a mistake.

not-an-aardvark commented on lib/rules/implicit-arrow-linebreak.js:
```diff
@@ -37,6 +43,130 @@ module.exports = {
+                case "VariableDeclarator":
+
+                    // find the grandparent, i.e. const, let, var
+                    return sourceCode.getFirstToken(node.parent.parent);
```
What will happen here if the parent `VariableDeclaration` has multiple `VariableDeclarator` children? For example:

```js
const foo = 1,
    bar = 2,
    baz = () => //comment
        qux
```

Will that result in the comment getting hoisted to the top of the statement, before `const`?

peanutenthusiast replied:
HI @not-an-aardvark, I added a couple test cases for when the parent `VariableDeclaration` has multiple `VariableDeclarator` children -- the comment should get hoisted to the top of the statement if the node is the first/only child, otherwise, the declarator itself.

platinumazure commented:
Thanks for contributing @peanutenthusiast and for your saintly patience as we processed this!
